lib/tasks/sync_hosts_with_teamdynamix.rake

```diff
@@ -11,26 +11,77 @@ desc <<-DESC.strip_heredoc.squish
-  task :sync_with_teamdynamix, [:hostnames] => :environment do |_task, args|
-    hostnames = Array.wrap(args.hostnames) + args.extras
-    if hostnames.present?
-      hosts = hostnames.collect { |hostname| Host.find_by(name: hostname) }
-    else
-      hosts = Host.all
-    end
-    error_count = 0
-
-    hosts.each do |host|
-      td_api = host.td_api
-      asset_id = host.teamdynamix_asset_id
-      action = asset_id && td_api.asset_exist?(asset_id) ? :update : :create
-      begin
-        td_api.send("#{action}_asset", host)
-      rescue StandardError => e
-        puts "\n\tS.No | hostname | action | result" if error_count.zero?
-        error_count += 1
-        puts "\t#{error_count} | #{host.name} | #{action} | Failed(#{e.message})"
+  task :sync_with_teamdynamix => :environment do |_task|
+    @td_api = TeamdynamixApi.instance
+    @errors = []
+    @hosts_synced = []
+
+    sync_existing_assets_to_hosts
+
+    create_assets_for_unmapped_hosts
+
+    print_summary
+  end
+
+  # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
+  def sync_existing_assets_to_hosts
+    # sending empty search param to get all the assets in TD at once
+    # Note: due to a limitation(possibly a bug) in TD search asset api
+    #       each and every search payload returns all the assets.
+    @teamdynamix_assets = @td_api.search_asset({})
+
+    @teamdynamix_assets.each do |asset|
+      asset_id = asset['ID']
+      # WHEN Asset is already mapped to a host
+      # THEN update the asset in TD as per current configuration
+      host = Host.find_by(teamdynamix_asset_id: asset_id)
+      if host.present?
+        @hosts_synced << host.id
+        @td_api.update_asset(host)
+        next
       end
+      hosts = Host.where('name IN (?)', [asset['Name'], asset['SerialNumber']])
+      # WHEN Asset does not have a matching host, THEN Do nothing
+      next if hosts.blank?
+      # WHEN Asset has more than one  matching host, THEN report error
+      if hosts.count > 1
+        @errors << "#{hosts.count} matching hosts found for asset with ID #{asset['ID']}"
+        next
+      end
+      # WHEN Asset has a uniquely matching host which is not yet synced
+      # THEN update the host with the asset ID
+      # AND update the asset in TD as per current configuration
+      host = hosts.first
+      @hosts_synced << host.id
+      host.teamdynamix_asset_id = asset_id
+      @errors << "failed to update host ##{host.id} for asset ID ##{asset_id}" unless host.save
+      @td_api.update_asset(host)
+    end
+  rescue StandardError => e
+    @errors << "component: syncing_assets_to_hosts, asset_id: #{asset['ID']},
+                asset_name: #{asset['Name']}, Error: #{e.message}"
+  end
+  # rubocop:enable Metrics/AbcSize, Metrics/MethodLength
+
+  # sync hosts that do not have assets created in Teamdynamix
+  def create_assets_for_unmapped_hosts
+    @hosts = Host.all
+    unmapped_hosts = @hosts.reject { |host| @hosts_synced.include?(host.id) }
```

Comment:
This reject can be replaced by a way more efficient db query

Reply:
sure